File: ui/src/gql/market.ts
import { gql } from "graphql-tag"

export const marketMk12StorageAskFragment = gql`
  fragment MarketMk12StorageAskAll on MarketMk12StorageAsk {
    spId
    price
    verifiedPrice
    minSize
    maxSize
    createdAt
    expiry
    sequence
  }
`

export const marketMk12PriceFilterFragment = gql`
  fragment MarketMk12PriceFilterAll on PriceFilter {
    name
    minDurationDays
    maxDurationDays
    minimumSize
    maximumSize
    price
    verified
  }`

export const marketMk12ClientFilterFragment = gql`
  fragment MarketMk12ClientFilterAll on ClientFilter {
    name
    active
    wallets
    peers
    pricingFilters
    maxDealsPerHour
    maxDealSizePerHour
    info
  }`

export const marketMk12DealFragment = gql`
  fragment MarketMk12DealCommon on MarketMk12Deal {
    uuid
    spId
    createdAt
    startEpoch
    endEpoch
    pieceCid
    pieceSize
  }
  fragment MarketMk12DealAll on MarketMk12Deal {
    ...MarketMk12DealCommon
    signedProposalCid
    proposalSignature
    proposal
    offline
    verified
    clientPeerId
    chainDealId
    publishCid
    fastRetrieval
    announceToIpni
    url
    urlHeaders
    error
    label
    proposalCid
  }
  `

export const GetMarketMk12StorageAsks = gql`
  query GetMarketMk12StorageAsks {
    marketMk12StorageAsks {
      ...MarketMk12StorageAskAll
    }
  }
  ${marketMk12StorageAskFragment}
  `

export const GetMarketMk12StorageAsk = gql`
  query GetMarketMk12StorageAsk($spId: Address!) {
    marketMk12StorageAsk(spId: $spId) {
      ...MarketMk12StorageAskAll
    }
  }
  ${marketMk12StorageAskFragment}
  `

export const UpdateMarketMk12StorageAsk = gql`
  mutation UpdateMarketMk12StorageAsk($input: MarketMk12StorageAskInput!) {
    updateMarketMk12StorageAsk(input: $input) {
      ...MarketMk12StorageAskAll
    }
  }
  ${marketMk12StorageAskFragment}
  `

export const marketBalanceFragment = gql`
  fragment MarketBalanceAll on MarketBalance {
    miner
    balance
    balances {
      address
      balance
    }
  }
`

export const GetMarketBalances = gql`
  query GetMarketBalances {
    marketBalances {
      ...MarketBalanceAll
    }
  }
  ${marketBalanceFragment}
`

export const AddMarketBalance = gql`
  mutation AddMarketBalance($miner: Address!, $amount: FIL!, $wallet: Address!) {
    marketAddBalance(miner: $miner, amount: $amount, wallet: $wallet) {
      ...MarketBalanceAll
    }
  }
  ${marketBalanceFragment}
`

export const GetMarketPriceFilters = gql`
  query GetMarketPriceFilters {
    makretPriceFilters {
      ...MarketMk12PriceFilterAll
    }
  }
  ${marketMk12PriceFilterFragment}
  `

export const AddMarketPriceFilter = gql`
  mutation AddMarketPriceFilter($input: PriceFilterInput!) {
    marketAddPriceFilter(input: $input)
  }
  `

export const UpdateMarketPriceFilter = gql`
  mutation UpdateMarketPriceFilter($input: PriceFilterInput!) {
    marketUpdatePriceFilter(input: $input) {
      ...MarketMk12PriceFilterAll
    }
  }
  ${marketMk12PriceFilterFragment}
  `

export const CheckMarketPriceFilter = gql`
  query CheckMarketPriceFilter($name: String!) {
    marketCheckPriceFilter(name: $name)
  }
  `

export const DeleteMarketPriceFilter = gql`
  mutation DeleteMarketPriceFilter($name: String!) {
    marketDeletePriceFilter(name: $name)
  }
`

export const GetMarketClientFilters = gql`
  query GetMarketClientFilters {
    marketClientFilters {
      ...MarketMk12ClientFilterAll
    }
  }
  ${marketMk12ClientFilterFragment}
  `

export const AddMarketClientFilter = gql`
  mutation AddMarketClientFilter($input: ClientFilterInput!) {
    marketAddClientFilter(input: $input)
  }
  `

export const UpdateMarketClientFilter = gql`
  mutation UpdateMarketClientFilter($input: ClientFilterInput!) {
    marketUpdateClientFilter(input: $input) {
      ...MarketMk12ClientFilterAll
    }
  }
  ${marketMk12ClientFilterFragment}
  `

export const CheckMarketClientFilter = gql`
  query CheckMarketClientFilter($name: String!) {
    marketCheckClientFilter(name: $name)
  }
  `

export const DeleteMarketClientFilter = gql`
  mutation DeleteMarketClientFilter($name: String!) {
    marketDeleteClientFilter(name: $name)
  }
  `

export const ToggleMarketClientFilter = gql`
  mutation ToggleMarketClientFilter($name: String!) {
    marketToggleClientFilter(name: $name)
  }
  `

export const GetMarketClientAllowList = gql`
  query GetMarketClientAllowList {
    marketAllowFilters {
      wallet
      status
    }
    marketDefaultFilterBehaviour {
      allowDealsFromUnknownClients
      isDealRejectedWhenCidGravityNotReachable
      isCidGravityEnabled {
        miner
        status
      }
    }
  }
  `

export const SetMarketClientAllow = gql`
  mutation SetMarketClientAllow($wallet: Address!, $status: Boolean!) {
    marketSetAllowFilter(wallet: $wallet, status: $status) {
      wallet
      status
    }
  }
  `

export const ToggleMarketClientAllow = gql`
  mutation ToggleMarketClientAllow($wallet: Address!) {
    marketToggleAllowFilter(wallet: $wallet)
  }
  `

export const DeleteMarketClientAllow = gql`
  mutation DeleteMarketClientAllow($wallet: Address!) {
    marketDeleteAllowFilter(wallet: $wallet)
  }
  `

export const GetMarketClientAllow = gql`
  query GetMarketClientAllow($wallet: Address!) {
    marketAllowFilter(wallet: $wallet) {
      wallet
      status
    }
  }
  `

export const GetMarketMk12Deals = gql`
  query GetMarketMk12Deals($filter: MarketMk12DealFilterInput! $offset: Int!, $limit: Int!) {
    marketMk12Deals(filter: $filter, offset: $offset, limit: $limit) {
      ...MarketMk12DealAll
    }
    marketMk12DealsCount(filter: $filter)
  }
  ${marketMk12DealFragment}
  `

File: graph/loaders/market.go
package loaders

import (
	"context"

	"github.com/filecoin-project/curio/web/api/webrpc"
	"github.com/web3tea/curio-dashboard/graph/model"
	"github.com/web3tea/curio-dashboard/types"
)

type MarketLoader interface {
	MarketMk12StorageAsks(ctx context.Context) ([]*model.MarketMk12StorageAsk, error)
	MarketMk12StorageAsk(ctx context.Context, spID uint64) (*model.MarketMk12StorageAsk, error)
	MarketMk12StorageAsksCount(ctx context.Context) (int, error)
	MarketMk12PriceFilter(ctx context.Context, name string) (*webrpc.PriceFilter, error)
	MarketMk12ClientFilter(ctx context.Context, name string) (*webrpc.ClientFilter, error)
	MarketAllowFilter(ctx context.Context, wallet types.Address) (*model.MarketAllowFilter, error)
	MarketMk12Deals(ctx context.Context, filter model.MarketMk12DealFilterInput, limit int, offset int) ([]*model.MarketMk12Deal, error)
	MarketMk12DealsCount(ctx context.Context, filter model.MarketMk12DealFilterInput) (int, error)
	MarketDealCountSummary(ctx context.Context) (*model.DealCountSummary, error)
}

type MarketLoaderImpl struct {
	loader *Loader
}

func NewMarketLoader(loader *Loader) MarketLoader {
	return &MarketLoaderImpl{loader}
}

func (l *MarketLoaderImpl) MarketMk12StorageAsks(ctx context.Context) ([]*model.MarketMk12StorageAsk, error) {
	var result []*model.MarketMk12StorageAsk

	err := l.loader.db.Select(ctx, &result, `SELECT
		sp_id,
		price,
		verified_price,
		min_size,
		max_size,
		created_at,
		expiry,
		sequence
		FROM market_mk12_storage_ask`)
	return result, err
}

func (l *MarketLoaderImpl) MarketMk12StorageAsk(ctx context.Context, spID uint64) (*model.MarketMk12StorageAsk, error) {
	var result model.MarketMk12StorageAsk

	err := l.loader.db.QueryRow(ctx, `SELECT
		sp_id,
		price,
		verified_price,
		min_size,
		max_size,
		created_at,
		expiry,
		sequence
		FROM market_mk12_storage_ask
		WHERE sp_id = $1`, spID).Scan(
		&result.SpID,
		&result.Price,
		&result.VerifiedPrice,
		&result.MinSize,
		&result.MaxSize,
		&result.CreatedAt,
		&result.Expiry,
		&result.Sequence,
	)
	return &result, err
}

func (l *MarketLoaderImpl) MarketMk12StorageAsksCount(ctx context.Context) (int, error) {
	var result int
	err := l.loader.db.QueryRow(ctx, `SELECT COUNT(*) FROM market_mk12_storage_ask`).Scan(&result)
	return result, err
}

func (l *MarketLoaderImpl) MarketMk12PriceFilter(ctx context.Context, name string) (*webrpc.PriceFilter, error) {
	var res []*webrpc.PriceFilter
	err := l.loader.db.Select(ctx, &res, `
	SELECT
		name,
		min_duration_days,
		max_duration_days,
		min_size,
		max_size,
		price,
		verified
	FROM market_mk12_pricing_filters
	WHERE name = $1`, name)
	if err != nil {
		return nil, err
	}
	if len(res) == 0 {
		return nil, ErrorNotFound
	}
	return res[0], nil
}

func (l *MarketLoaderImpl) MarketMk12ClientFilter(ctx context.Context, name string) (*webrpc.ClientFilter, error) {
	var res []*webrpc.ClientFilter
	err := l.loader.db.Select(ctx, &res, `
	SELECT
		name,
		active,
		wallets,
		peer_ids,
		pricing_filters,
		max_deals_per_hour,
		max_deal_size_per_hour,
		additional_info
	FROM market_mk12_client_filters
	WHERE name = $1`, name)
	if err != nil {
		return nil, err
	}
	if len(res) == 0 {
		return nil, ErrorNotFound
	}
	return res[0], nil
}

func (l *MarketLoaderImpl) MarketAllowFilter(ctx context.Context, wallet types.Address) (*model.MarketAllowFilter, error) {
	var res []*model.MarketAllowFilter
	err := l.loader.db.Select(ctx, &res, `
	SELECT
		wallet,
		status
	FROM market_allow_list
	WHERE wallet = $1`, wallet)
	if err != nil {
		return nil, err
	}
	if len(res) == 0 {
		return nil, ErrorNotFound
	}
	return res[0], nil
}

func (l *MarketLoaderImpl) MarketMk12Deals(ctx context.Context, filter model.MarketMk12DealFilterInput, limit int, offset int) ([]*model.MarketMk12Deal, error) {
	var result []*model.MarketMk12Deal
	err := l.loader.db.Select(ctx, &result, `SELECT
            uuid,
            sp_id,
            created_at,
            signed_proposal_cid,
            proposal_signature,
            proposal,
            offline,
            verified,
            start_epoch,
            end_epoch,
            client_peer_id,
            chain_deal_id,
            publish_cid,
            piece_cid,
            piece_size,
            fast_retrieval,
            announce_to_ipni,
            url,
            url_headers,
            error,
            label,
            proposal_cid
        FROM market_mk12_deals
        WHERE 1=1
        AND ($1::int8 IS NULL OR sp_id = $1)
        AND ($2::text IS NULL OR uuid = $2)
        AND ($3::text IS NULL OR signed_proposal_cid = $3)
        AND ($4::text IS NULL OR proposal_cid = $4)
        AND ($5::text IS NULL OR piece_cid = $5)
        LIMIT $6 OFFSET $7
    `,
		filter.SpID,
		filter.UUID,
		filter.SignedProposalCid,
		filter.ProposalCid,
		filter.PieceCid,
		limit,
		offset,
	)
	if err != nil {
		return nil, err
	}
	return result, nil
}

func (l *MarketLoaderImpl) MarketMk12DealsCount(ctx context.Context, filter model.MarketMk12DealFilterInput) (int, error) {
	var result int
	err := l.loader.db.QueryRow(ctx, `SELECT COUNT(*) FROM market_mk12_deals
		WHERE 1=1
		AND ($1::int8 IS NULL OR sp_id = $1)
		AND ($2::text IS NULL OR uuid = $2)
		AND ($3::text IS NULL OR signed_proposal_cid = $3)
		AND ($4::text IS NULL OR proposal_cid = $4)
		AND ($5::text IS NULL OR piece_cid = $5)
	`,
		filter.SpID,
		filter.UUID,
		filter.SignedProposalCid,
		filter.ProposalCid,
		filter.PieceCid,
	).Scan(&result)
	return result, err
}

func (l *MarketLoaderImpl) MarketDealCountSummary(ctx context.Context) (*model.DealCountSummary, error) {
	summary := &model.DealCountSummary{}

	var boostCount int
	err := l.loader.db.QueryRow(ctx, `SELECT COUNT(*) FROM market_mk12_deals`).Scan(&boostCount)
	if err != nil {
		return nil, err
	}
	summary.Boost = boostCount

	var directCount int
	err = l.loader.db.QueryRow(ctx, `SELECT COUNT(*) FROM market_direct_deals`).Scan(&directCount)
	if err != nil {
		return nil, err
	}
	summary.Direct = directCount

	var legacyCount int
	err = l.loader.db.QueryRow(ctx, `SELECT COUNT(*) FROM market_legacy_deals`).Scan(&legacyCount)
	if err != nil {
		return nil, err
	}
	summary.Legacy = legacyCount

	return summary, nil
}
